ruby/lib/endpoints/clusters.rb

```diff
@@ -0,0 +1,104 @@
+#!/usr/bin/env ruby
+# encoding: UTF-8
+#
+# Copyright © 2016 Cask Data, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+require_relative 'types'
+
+module CmApi
+  module Endpoints
+    # Module for Cluster methods and types
+    module Clusters
+      include ::CmApi::Endpoints::Types
+
+      CLUSTERS_PATH = '/clusters'.freeze
+
+      def create_cluster(name, version = nil, fullVersion = nil)
+        if version.nil? && fullVersion.nil?
+          raise "Either 'version' or 'fullVersion' must be specified"
+        end
+        if !fullVersion.nil?
+          api_version = 6
+          version = nil
+        else
+          api_version = 1
+        end
+
+        apicluster = ApiCluster.new(self, name, version, fullVersion)
+        call_resource(method(:post), CLUSTERS_PATH, ApiCluster, true, [apicluster], nil, api_version)[0]
+      end
+
+      def get_cluster(name)
+        call_resource(method(:get), "#{CLUSTERS_PATH}/#{name}", ApiCluster)
+      end
+
+      def get_all_clusters(view = nil)
+        call_resource(method(:get), CLUSTERS_PATH, ApiCluster, true, nil, view && { 'view' => view } || nil)
+      end
+
+      def delete_cluster(name)
+        call_resource(method(:delete), "#{CLUSTERS_PATH}/#{name}", ApiCluster)
+      end
+
+      # Model for a cluster
+      class ApiCluster < BaseApiResource
+        @_ATTRIBUTES = {
+          'name' => nil,
+          'displayName' => nil,
+          'clusterUrl' => nil,
+          'version' => nil,
+          'fullVersion' => nil,
+          'hostsUrl' => ROAttr.new,
+          'maintenanceMode' => ROAttr.new,
+          'maintenanceOwners' => ROAttr.new,
+          'entityStatus' => ROAttr.new
+        }
+
+        def initialize(resource_root, name = nil, version = nil, fullVersion = nil)
+          # possible alternative to generate the hash argument dynamically, similar to python locals():
+          #  method(__method__).parameters.map { |arg| arg[1] }.inject({}) { |h, a| h[a] = eval a.to_s; h}
+          super(resource_root, { name: name, version: version, fullVersion: fullVersion })
+        end
+
+        def to_s
```

Comment:
actually ruby has both ``to_s`` and ``to_str`` and they are different... after reading some docs, the bug here is that i need to convert the other ``to_str`` to ``to_s``.  basically, ``to_s`` is the string representation of an object, which every object should have, and is relevant here.  ``to_str`` is only for when an object really is a string and can have string operations performed on them.